File: backend/src/app/domain/pagination.py
import math
from dataclasses import dataclass
from enum import Enum
from typing import Generic, TypeVar

T = TypeVar("T")

MIN_PAGE = 1
MIN_PAGE_SIZE = 1
MAX_PAGE_SIZE = 100
DEFAULT_PAGE_SIZE = 10


class SortDirection(str, Enum):
    ASC = "asc"
    DESC = "desc"


@dataclass(frozen=True)
class Page(Generic[T]):
    items: list[T]
    total: int
    page: int
    page_size: int

    @property
    def total_pages(self) -> int:
        return math.ceil(self.total / self.page_size) if self.page_size > 0 else 0

    @property
    def offset(self) -> int:
        return (self.page - 1) * self.page_size

File: frontend/src/domain/pagination.ts
export type SortDirection = 'asc' | 'desc';

export type UserSortField = 'id' | 'name' | 'username' | 'email';

export interface SortState<TField extends string> {
  field: TField;
  direction: SortDirection;
}

export interface Page<TItem> {
  items: TItem[];
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
}

export function nextSortState<TField extends string>(
  current: SortState<TField>,
  field: TField,
): SortState<TField> {
  if (current.field !== field) {
    return { field, direction: 'asc' };
  }
  return { field, direction: current.direction === 'asc' ? 'desc' : 'asc' };
}
